src/modules/flow/freegeoip/freegeoip.c

```diff
@@ -0,0 +1,239 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "freegeoip-gen.h"
+#include "sol-flow-internal.h"
+
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+
+#include <sol-http-client.h>
+#include <sol-json.h>
+#include <sol-mainloop.h>
+#include <sol-util.h>
+
+struct freegeoip_data {
+    struct sol_flow_node *node;
+    char *endpoint;
+};
+
+#define RESPONSE_CHECK_API(response_, mdata_) \
+    do { \
+        if (unlikely(!response_)) { \
+            sol_flow_send_error_packet(mdata_->node, EINVAL, \
+                "Error while reaching Freegeoip"); \
+            return; \
+        } \
+        if (unlikely(response_->api_version != SOL_HTTP_RESPONSE_API_VERSION)) { \
+            SOL_ERR("Unexpected API version (response is %u, expected %u)", \
+                response->api_version, SOL_HTTP_RESPONSE_API_VERSION); \
+            return; \
+        } \
+    } while (0)
+
+enum fields {
+    FIELD_IP = 1<<0,
+    FIELD_COUNTRY_CODE = 1<<1,
+    FIELD_COUNTRY_NAME = 1<<2,
+    FIELD_REGION_CODE = 1<<3,
+    FIELD_CITY_NAME = 1<<4,
+    FIELD_ZIP_CODE = 1<<5,
+    FIELD_TIMEZONE = 1<<6,
+    FIELD_LATITUDE = 1<<7,
+    FIELD_LONGITUDE = 1<<8,
+    FIELD_ALL = 0x1ff
+};
+
+static bool
+json_token_to_double(const struct sol_json_token *token, double *output)
+{
+    char *str = strndupa(token->start, token->end - token->start);
+    char *endptr;
+    double value;
+
+    errno = 0;
+    value = strtod(str, &endptr);
+    if (errno != 0)
+        return false;
+    if (!endptr || *endptr)
+        return false;
+    *output = value;
+    return true;
+}
+
+static void
+freegeoip_query_finished(void *data, struct sol_http_response *response)
+{
+    struct freegeoip_data *mdata = data;
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+    enum fields fields = FIELD_ALL;
+    double latitude = 0.0f;
+    double longitude = 0.0f;
+
+    RESPONSE_CHECK_API(response, mdata);
+
+    if (!response->content.used) {
+        SOL_WRN("Empty response from FreeGeoIP");
+        return;
+    }
+
+    if (response->response_code != 200) {
+        sol_flow_send_error_packet(mdata->node, EINVAL,
+            "FreeGeoIP returned an unknown response code: %d",
+            response->response_code);
+        return;
+    }
+
+#define FIELD(json_field_, flow_field_) \
+    do { \
+        if (fields & FIELD_ ## flow_field_ && sol_json_token_str_eq(&key, \
+                json_field_, strlen(json_field_))) { \
+            fields &= ~(FIELD_ ## flow_field_); \
+            sol_flow_send_string_packet(mdata->node, \
+                SOL_FLOW_NODE_TYPE_LOCATION_FREEGEOIP__OUT__ ## flow_field_, \
+                strndupa(value.start + 1, value.end - value.start - 2)); \
```

Comment:
we should have a way to send a sol_str_slice... then there is no need to strndupa().